pkg/gnoi/file/file.go

```diff
@@ -0,0 +1,163 @@
+	if url == "" {
+		return nil, status.Error(codes.InvalidArgument, "remote download path (URL) cannot be empty")
+	}
+
+	// Container path translation: prepend /mnt/host to access host filesystem
+	// Only apply if /mnt/host exists (running in container) and path doesn't already have it
+	translatedPath := translatePathForContainer(localPath)
+
+	// Download file from URL
+	if err := download.DownloadHTTP(ctx, url, translatedPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "download failed: %v", err)
+	}
+
+	// Calculate MD5 hash of downloaded file
+	hashBytes, err := hash.CalculateMD5(translatedPath)
+	if err != nil {
+		// Clean up the downloaded file since we can't verify it
+		os.Remove(translatedPath)
+		return nil, status.Errorf(codes.Internal, "hash calculation failed: %v", err)
+	}
+
+	// Build response with MD5 hash
+	return &gnoi_file_pb.TransferToRemoteResponse{
+		Hash: &types.HashType{
+			Method: types.HashType_MD5,
+			Hash:   hashBytes,
+		},
+	}, nil
+}
+
+// translatePathForContainer handles path translation for container environments.
+// If the code is running in a container with /mnt/host mount (host filesystem access),
+// it prepends /mnt/host to the path. This follows the same pattern as the diskspace package.
+//
+// Example:
+//   - Input: "/tmp/firmware.bin"
+//   - Running in container: "/mnt/host/tmp/firmware.bin"
+//   - Running on host: "/tmp/firmware.bin"
+func translatePathForContainer(path string) string {
+	// Clean the path first
+	cleanPath := filepath.Clean(path)
+
+	// If path already starts with /mnt/host, don't double-prefix
+	if strings.HasPrefix(cleanPath, "/mnt/host") {
+		return cleanPath
+	}
+
+	// Check if /mnt/host exists (indicates we're running in a container)
+	if _, err := os.Stat("/mnt/host"); err == nil {
+		return "/mnt/host" + cleanPath
+	}
+
+	// Not in container, return original path
+	return cleanPath
+}
+
+// validatePath checks if the requested path is within allowed directories.
+// This prevents security issues like overwriting critical system files.
+//
+// Allowed directories for SONiC devices:
+//   - /tmp/               - Temporary files, firmware images
+//   - /host/              - Persistent storage (firmware, binaries, config)
+//   - /var/log/           - Log files
+//
+// Rejected paths:
+//   - /etc/, /boot/, /usr/, /bin/, /sbin/ - Critical system directories
+//   - Relative paths or paths with .. traversal
+func validatePath(path string) error {
+	// Clean the path to resolve . and .. components
+	cleanPath := filepath.Clean(path)
+
+	// Must be absolute path
+	if !filepath.IsAbs(cleanPath) {
+		return fmt.Errorf("path must be absolute, got: %s", path)
+	}
+
+	// Check if path contains .. after cleaning (path traversal attempt)
+	if strings.Contains(cleanPath, "..") {
+		return fmt.Errorf("path traversal not allowed: %s", path)
+	}
+
+	// Whitelist of allowed directory prefixes
+	allowedPrefixes := []string{
+		"/tmp/",
+		"/host/",
```

Comment:
`/host/image-.../rw/` directory has all of the files that have been modified compared to the original squashfs file for this image. In the best case, it allows someone to create a file on the host anywhere. In the worst case, it can cause a kernel panic as overlayfs may get confused about the state of the filesystem.

Reply:
`/host` also has some sensitive directories like `/host/grub/grub.cfg`; if that is modified, then rebooting the device may put it in an unexpected state.